activation/activation_test.go

```diff
@@ -1487,3 +1485,99 @@ func TestFindFullyValidHighTickAtx(t *testing.T) {
+// for the current epoch). This behavior is mocked by mocking the response of the
+// proof validator.
+func Test_Builder_RegenerateInitialPost(t *testing.T) {
+	var (
+		tab           = newTestBuilder(t, 1)
+		sig           = maps.Values(tab.signers)[0]
+		genesis       = time.Now()
+		ctx, cancel   = context.WithCancel(context.Background())
+		commitmentATX = types.RandomATXID()
+		nonce         = types.VRFPostIndex(rand.Uint64())
+		numUnits      = uint32(12)
+		initialPost   = &types.Post{
+			Nonce:   rand.Uint32(),
+			Indices: types.RandomBytes(10),
+			Pow:     rand.Uint64(),
+		}
+	)
+
+	postV2Calls := 0
+	tab.mValidator.EXPECT().
+		PostV2(gomock.Any(), sig.NodeID(), commitmentATX, initialPost, shared.ZeroChallenge, numUnits).DoAndReturn(func(_ context.Context, _ types.NodeID, _ types.ATXID, _ *types.Post, _ []byte, _ uint32, _ ...validatorOption) error {
+		switch postV2Calls {
+		case 0, 1, 3:
+			postV2Calls++
+			return nil
+		case 2:
+			postV2Calls++
+			return nil
+		}
+		panic("unexpected test branch")
+	}).Times(4)
+
+	tab.mclock.EXPECT().CurrentLayer().Return(types.LayerID(0)).AnyTimes()
+	tab.mclock.EXPECT().AwaitLayer(gomock.Any()).DoAndReturn(func(id types.LayerID) <-chan struct{} {
+		// this is our way to end the test - the code will wait on AwaitLayer in PublishActivationTx while
+		// waiting for the publication epoch. otherwise the `run` method will keep on looping to broadcast
+		// future atxs
+		cancel()
+		return make(chan struct{})
+	}).AnyTimes()
+	tab.mclock.EXPECT().LayerToTime(gomock.Any()).DoAndReturn(func(lid types.LayerID) time.Time {
+		// layer duration is 10ms to speed up test
+		return genesis.Add(time.Duration(lid) * 20 * time.Millisecond)
+	}).AnyTimes()
+	tab.mnipost.EXPECT().ResetState(sig.NodeID()).Return(nil).Times(1)
+
+	nipostCalls := 0
+	tab.mnipost.EXPECT().
+		BuildNIPost(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+		DoAndReturn(func(context.Context, *signing.EdSigner, types.EpochID, types.Hash32, *types.NIPostChallenge) (*nipost.NIPostState, error) {
+			switch nipostCalls {
+			case 0:
+				nipostCalls++
+				return nil, ErrInvalidInitialPost
+			case 1:
+				nipostCalls++
+				return &nipost.NIPostState{}, nil
+			default:
+				panic("unexpected branch")
+			}
+
+		}).AnyTimes()
+
+	proofCalls := 0
+	tab.mnipost.EXPECT().Proof(gomock.Any(), sig.NodeID(), shared.ZeroChallenge, nil).DoAndReturn(
+		func(context.Context, types.NodeID, []byte, *types.NIPostChallenge) (*types.Post, *types.PostInfo, error) {
+			switch proofCalls {
+			case 0, 1, 2:
+				proofCalls++
+				return initialPost,
+					&types.PostInfo{
+						NodeID:        sig.NodeID(),
+						CommitmentATX: commitmentATX,
+						Nonce:         &nonce,
+
+						NumUnits:      numUnits,
+						LabelsPerUnit: DefaultPostConfig().LabelsPerUnit,
+					},
+					nil
+			default:
+				panic("unexpected branch")
+			}
+		}).AnyTimes()
```

Comment:
Same here:

```suggestion
	tab.mnipost.EXPECT().Proof(gomock.Any(), sig.NodeID(), shared.ZeroChallenge, nil).Return(
		initialPost,
		&types.PostInfo{
			NodeID:        sig.NodeID(),
			CommitmentATX: commitmentATX,
			Nonce:         &nonce,

			NumUnits:      numUnits,
			LabelsPerUnit: DefaultPostConfig().LabelsPerUnit,
		}, nil).Times(2)
```

This method is actually only called twice during the test, not 3 times.